File: Python/0003_longest_substring_without_repeating_characters/0003_longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        '''
        A Substring without repeating characters,
        Using two pointers,(start and i, i is current traversed char),
        No repeating char between two pointers,
        So, the max length between two pointers is the answer.
        When find the repeating char,
        the start pointer will move to the position + 1 of appeared char where found last time,
        continued traversed until i == len(s).
        '''
        mapping = [-1] * 256
        maximum = 0
        start = -1
        for i in range(len(s)):
            if mapping[ord(s[i])] > start:
                start = mapping[ord(s[i])]
            mapping[ord(s[i])] = i
            maximum = max(maximum, i - start)
        return maximum

File: Java/0003_longest_substring_without_repeating_characters/0003_longest_substring_without_repeating_characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] map = new int[256];
        Arrays.fill(map, -1);
        int maximum = 0;
        int start = -1;
        for (int i = 0; i < s.length(); i++) {
            if (map[s.charAt(i)] > start)
                start = map[s.charAt(i)];
            map[s.charAt(i)] = i;
            maximum = Math.max(maximum, i - start);
        }
        return maximum;
    }
}
/*
A Substring without repeating characters,
Using two pointers,(start and i, i is current traversed char),
No repeating char between two pointers,
So, the max length between two pointers is the answer.
When find the repeating char,
the start pointer will move to the position + 1 of appeared char where found last time,
continued traversed until i == len(s).
*/
